src/sleep.rs

```diff
@@ -0,0 +1,376 @@
+    }
+
+    pub fn apply(&self) -> Result<(), EspError> {
+        esp!(unsafe { esp_sleep_enable_timer_wakeup(self.duration.as_micros() as u64) })?;
+        Ok(())
+    }
+}
+
+/// Will wake up the CPU based on changes in RTC GPIO pins. Is available for both light and
+/// deep sleep. However, for light sleep, the GpioWakeup struct offers more flexibility.
+/// Pullup and pulldown can be enabled for each pin. This settings will be applied when
+/// deep sleep is enabled.
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub struct RtcWakeup<'a> {
+    pub pins: &'a [RtcWakeupPin],
+    pub wake_level: RtcWakeLevel,
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<'a> RtcWakeup<'a> {
+    pub fn new(pins: &'a [RtcWakeupPin], wake_level: RtcWakeLevel) -> Self {
+        Self { pins, wake_level }
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<'a> RtcWakeup<'a> {
+    fn mask(&self) -> u64 {
+        let mut m: u64 = 0;
+        for pin in self.pins.iter() {
+            m |= 1 << pin.pin.pin();
+        }
+        m
+    }
+
+    pub fn apply(&self) -> Result<(), EspError> {
+        #[cfg(any(esp32, esp32s3))]
+        for pin in self.pins {
+            if pin.pullup || pin.pulldown {
+                esp!(unsafe {
+                    esp_sleep_pd_config(
+                        esp_sleep_pd_domain_t_ESP_PD_DOMAIN_RTC_PERIPH,
+                        esp_sleep_pd_option_t_ESP_PD_OPTION_ON,
+                    )
+                })?;
+
+                if pin.pullup {
+                    esp!(unsafe { rtc_gpio_pullup_en(pin.pin.pin()) })?;
+                }
+                if pin.pulldown {
+                    esp!(unsafe { rtc_gpio_pulldown_en(pin.pin.pin()) })?;
+                }
+            }
+        }
+
+        if !self.pins.is_empty() {
+            /* Do not use ext0, as ext1 with one pin does the same */
+            esp!(unsafe { esp_sleep_enable_ext1_wakeup(self.mask(), self.wake_level.mode()) })?;
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<'a> fmt::Debug for RtcWakeup<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "RtcWakeup {{ pins: [")?;
+
+        for pin in self.pins.iter() {
+            write!(f, "{:?} ", pin.pin.pin())?;
+        }
+        write!(
+            f,
+            "] mask: {:#x} wake_level: {:?} }}",
+            self.mask(),
+            self.wake_level
+        )
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+#[derive(Debug)]
+pub enum RtcWakeLevel {
+    AllLow,
+    AnyHigh,
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl RtcWakeLevel {
+    pub fn mode(&self) -> u32 {
+        match self {
+            RtcWakeLevel::AllLow => esp_sleep_ext1_wakeup_mode_t_ESP_EXT1_WAKEUP_ALL_LOW,
+            RtcWakeLevel::AnyHigh => esp_sleep_ext1_wakeup_mode_t_ESP_EXT1_WAKEUP_ANY_HIGH,
+        }
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub struct RtcWakeupPin {
+    pub pin: AnyInputPin, // this should use RtcInput somehow
```

Comment:
Note that the above is also applicable for `GpioWakeupPin`.

And just like the case for `GpioWakeupPin` and UART, same question here (also see below): should the RTC pin be configured, and in RTC mode, for the wakeup to happen or not? If the former, we need to pass here a `PinDriver` in `Input` mode rather than the raw RTC pin peripheral.